elastic-agent-client.proto

```diff
@@ -112,15 +114,301 @@ message ActionResponse {
+    string unit_id = 2;
+    // Type of transaction to start.
+    StoreTxnType type = 3;
+}
+
+// Response for a started transaction.
+message StoreBeginTxnResponse {
+    // Transaction ID.
+    string id = 1;
+}
+
+// Gets a key from the store.
+message StoreGetKeyRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+    // Name of the key.
+    string name = 3;
+}
+
+// Response of the retrieved key.
+message StoreGetKeyResponse {
+    // Status result of the get.
+    enum Status {
+        // Action was successful.
+        FOUND = 0;
+        // Action has failed.
+        NOT_FOUND = 1;
+    }
+    Status status = 1;
+    // Value when `FOUND`.
+    bytes value = 2;
+    // Metadata when `FOUND`.
+    bytes metadata = 3;
+}
+
+// Sets a key into the store.
+//
+// `txn_id` must be an ID of a transaction that was started with `READ_WRITE`.
+message StoreSetKeyRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+    // Name of the key.
+    string name = 3;
+    // Value of the key.
+    bytes value = 4;
+    // Metadata of the key.
+    bytes metadata = 5;
+    // TTL of the key (in milliseconds)
+    uint64 ttl = 6;
+}
+
+// Response from `SetKey`.
+message StoreSetKeyResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Deletes a key in the store.
+//
+// `txn_id` must be an ID of a transaction that was started with `READ_WRITE`.
+//
+// Does not error in the case that a key does not exist.
+message StoreDeleteKeyRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+    // Name of the key.
+    string name = 3;
+}
+
+// Response from `DeleteKey`.
+message StoreDeleteKeyResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Commits the transaction in the store.
+//
+// Upon error the whole transaction is discarded so no need to call discard after error.
+message StoreCommitRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+}
+
+// Response from `CommitTxn`.
+message StoreCommitTxnResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+// Discards the transaction in the store.
+message StoreDiscardRequest {
+    // Token that is used to uniquely identify the connection to the Elastic Agent.
+    string token = 1;
+    // Transaction ID.
+    string txn_id = 2;
+}
+
+// Response from `DiscardTxn`.
+message StoreDiscardTxnResponse {
+    // Empty at the moment, defined for possibility of adding future return values.
+}
+
+service ElasticAgentStore {
```

Comment:
Why is this store needed? Who are the users of this?

Reply:
This is for all users, to ease their implementation. Hopefully filestreams and other inputs will transition to using the key-value store. The benefit is the life-cycle of the state is handled by the Elastic Agent.

Elastic Agent will clean up state for units, as well manage the state across upgrade of versions. No need for the inputs to worry about writing state and managing the life-cycle of it.

Reply:
On the surface-level it sounds great. But after thinking about it, I am on the fence. If e.g. `filestream` uses agent to store state information, it means that we cannot really test `filestream` without either starting an Agent or mocking it. I am afraid this adds unnecessary overhead to development of inputs. During testing I would like to avoid starting an Agent whenever it is possible. In e2e tests it is fine, but for unit and integration tests it sounds too much. But let's see.